### ISSUE TITLE
NuGet.Server should come along with symbol serving support

### ISSUE THREAD

**Issue by deep-outcome** [NuGet/NuGetGallery#8482]
### Situation
> Since NuGet.Server does not come with a symbol server, makes sense to …

[NuGet.Server](https://www.nuget.org/packages/NuGet.Server/) lightweight NuGet server cannot serve as a symbol server. 👎🏻

### Request

Include a symbol server in the NuGet.Server. 👍🏻

**Comment by joelverhagen** [comment]
Hello @Uzivatel919, thanks for taking time to open this feature request. Currently, NuGet.Server only supports NuGet packages and not symbol packages or serving symbol (.pdb) files, as you have stated.

Currently NuGet.Server is a lower priority for our team. NuGet.Server only supports the V2 protocol which is in maintenance mode and not getting new features. Most of our team's energies are focusing on NuGet.org itself.

It is technically possible to support serving symbols in NuGet.Server but it would likely require help from the community to make this feature request happen, in the form of writing a small design and opening one or a few pull requests. For reference, in the past a community member made a huge contribution and helped us move NuGet.Server to ASP.NET to support multiple feeds in a single process.

In the meantime I'd recommend using one of the [3rd party package sources mentioned in our docs](https://docs.microsoft.com/en-us/nuget/hosting-packages/overview) that also support symbols. You may need to check docs to confirm which servers supports symbols.

Off the top of my head, perhaps you can look into:

- [BaGet](https://github.com/loic-sharma/BaGet) - similar to NuGet.Server in that it can run in a local server
- [Sleet](https://github.com/emgarten/sleet) - serverless, if you want to avoid managing the web server yourself

Another option that some folks leverage is to bundle the PDB files into the main NuGet package. See [Alternative PDB distribution](https://github.com/dotnet/sourcelink#alternative-pdb-distribution) on the SourceLink for more information about some pros and cons on using .snupkg. There's also some interesting information on this [Azure Artifacts feature request page](https://developercommunity.visualstudio.com/t/add-snupkg-support-to-azure-devops-artifacts/657354) for supporting .snupkg on Azure Artifacts.

**Comment by Uzivatel919** [comment]
Hello @joelverhagen, thanks for apposite answer. I will stuck with `<AllowedOutputExtensionsInPackageBuildOutputFolder>$(AllowedOutputExtensionsInPackageBuildOutputFolder);.pdb</AllowedOutputExtensionsInPackageBuildOutputFolder>` for this time.

Are there any starting points if I would be for starting one or few PRs?

**Comment by joelverhagen** [comment]
If you wanted to contribute this feature to NuGet.Server, I would follow the following steps:

1. Draft a short design document to describe your changes and how they would work.
    - Instructions for a proposal: https://github.com/NuGet/Home/blob/dev/meta/README.md
    - This step should be done before you open the PRs for review but you can prototype if you want to help write the design.
    - Some questions in my mind:
       - How does the client discover the symbol push endpoint? In V3, there is a [resource](https://docs.microsoft.com/en-us/nuget/api/symbol-package-publish-resource) for service discovery. There is no similar flow in V2 and I'm not even sure how NuGet push behaves with pushing .snupkg packages to a V2 feed.
       - What URL pattern would the symbols be served from? The base URL of this pattern would be what you configure into the VS debugger symbol server URLs.
1. Break the work up into phases so that the PRs are not too big (this helps us to review in a timely manner).
    - For example, PRs could be:
       1. Add C# classes that accept a .snupkg, pull out PDBs, and save them to disk in a structured format for serving future requests. Something like [`IServerPackageRepository`](https://github.com/NuGet/NuGet.Server/blob/main/src/NuGet.Server.Core/Infrastructure/IServerPackageRepository.cs) but for .snupkg files. You'll need to validate that the PDB matches the DLL with code like what NuGet.org uses in our backend: [`SymbolValidatorService`](https://github.com/NuGet/NuGet.Jobs/blob/main/src/Validation.Symbols/SymbolsValidatorService.cs).
       2. Add the REST API for reading symbols. This is not a NuGet protocol but a PDB/debugger related protocol. I don't know the details of it but both BaGet and Sleet are open source so you can probably learn what you need there. Maybe @loic-sharma or @emgarten have references on the symbol server protocol so you can learn what the REST API should look like.
       3. Add the REST API for pushing symbols. The protocol to follow is here: https://docs.microsoft.com/en-us/nuget/api/symbol-package-publish-resource. The .snupkg push endpoint is a lot like the existing .nupkg endpoint but it's a different URL (NuGet.org has `/api/v2/symbolpackage` instead of `/api/v2/package`) and requires a package with `SymbolsPackage` package type ([more details on .snupkg](https://docs.microsoft.com/en-us/nuget/create-packages/symbol-packages-snupkg#symbol-package-structure)). In the NuGet.org case we validate the .snupkg matches the .nupkg asynchronously (in the background, after push) but in NuGet.Server you could validate it before the push completes and return an HTTP 400 if a problem is found.
 1. Open the PRs in sequence and get reviews 😄.

**Comment by loic-sharma** [comment]
> This is not a NuGet protocol but a PDB/debugger related protocol. I don't know the details of it but both BaGet and Sleet are open source so you can probably learn what you need there. Maybe @loic-sharma or @emgarten have references on the symbol server protocol so you can learn what the REST API should look like.

Here are some helpful resources:

* Symbol server and PDB docs
    * https://github.com/dotnet/symstore/blob/main/docs/specs/Simple_Symbol_Query_Protocol.md
    * https://github.com/dotnet/symstore/blob/main/docs/specs/SSQP_Key_Conventions.md
    * https://github.com/dotnet/core/blob/main/Documentation/diagnostics/portable_pdb.md
    * https://github.com/dotnet/runtime/blob/main/docs/design/specs/PortablePdb-Metadata.md
* Interesting code
    * BaGet
        * [`SymbolController.Upload`](https://github.com/loic-sharma/BaGet/blob/5fbcc5e001abe4c9b3030a4412ee7ab4184c6e75/src/BaGet.Hosting/Controllers/SymbolController.cs#L34)
        * [`SymbolController.Get](https://github.com/loic-sharma/BaGet/blob/5fbcc5e001abe4c9b3030a4412ee7ab4184c6e75/src/BaGet.Hosting/Controllers/SymbolController.cs#L78)
        * [`SymbolIndexingService.IndexAsync`](https://github.com/loic-sharma/BaGet/blob/5fbcc5e001abe4c9b3030a4412ee7ab4184c6e75/src/BaGet.Core/Indexing/SymbolIndexingService.cs#L41)
    * Sleet
        * [Symbols.GetAssembliesAsync](https://github.com/emgarten/Sleet/blob/5d483588c1f3e3f09426ffad1b45020ec09ec1d5/src/SleetLib/Services/Symbols.cs#L286)
        * [SymbolsUtility](https://github.com/emgarten/Sleet/blob/5d483588c1f3e3f09426ffad1b45020ec09ec1d5/src/SleetLib/Utility/SymbolsUtility.cs)
